wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs

```diff
@@ -0,0 +1,155 @@
+instance Arbitrary CreateAddressError where
+    arbitrary = oneof []
+
+instance Buildable CreateAddressError where
+    build (CreateAddressUnknownHdAccount uAccount) =
+        bprint ("CreateAddressUnknownHdAccount " % F.build) uAccount
+    build (CreateAddressKeystoreNotFound accId) =
+        bprint ("CreateAddressKeystoreNotFound " % F.build) accId
+    build (CreateAddressHdRndGenerationFailed hdAcc) =
+        bprint ("CreateAddressHdRndGenerationFailed " % F.build) hdAcc
+    build (CreateAddressHdRndAddressSpaceSaturated hdAcc) =
+        bprint ("CreateAddressHdRndAddressSpaceSaturated " % F.build) hdAcc
+
+-- | Creates a new 'Address' for the input account.
+createAddress :: PassPhrase
+              -- ^ The 'Passphrase' (a.k.a the \"Spending Password\").
+              -> AccountId
+              -- ^ An abstract notion of an 'Account' identifier
+              -> PassiveWallet
+              -> IO (Either CreateAddressError Address)
+createAddress spendingPassword accId pw = do
+    let keystore = pw ^. walletKeystore
+    case accId of
+         -- \"Standard\" HD random derivation. The strategy is as follows:
+         --
+         -- 1. Generate the Address' @index@ and @HdAddress@ structure outside
+         --    of an atomic acid-state transaction. This could lead to data
+         --    races in the sense that an index is picked and such index
+         --    is already claimed, but if this happens we simply try again.
+         -- 2. Perform the actual creation of the 'HdAddress' as an atomic
+         --    transaction in acid-state.
+         --
+         -- The reason why we do this is because:
+         -- 1. We cannot do IO (thus index derivation) in an acid-state
+         --    transaction
+         -- 2. In order to create an 'HdAddress' we need a proper 'Address',
+         -- but this cannot be derived with having access to the
+         -- 'EncryptedSecretKey' and the 'PassPhrase', and we do not want
+         -- these exposed in the acid-state transaction log.
+         (AccountIdHdRnd hdAccId) -> do
+             mbEsk <- Keystore.lookup (WalletIdHdRnd (hdAccId ^. hdAccountIdParent))
+                                      keystore
+             case mbEsk of
+                  Nothing  -> return (Left $ CreateAddressKeystoreNotFound accId)
+                  Just esk -> createHdRndAddress spendingPassword esk hdAccId pw
+
+-- | Creates a new 'Address' using the random HD derivation under the hood.
+-- Being this an operation bound not only by the number of available derivation
+-- indexes \"left\" in the account, some form of short-circuiting is necessary.
+-- Currently, the algorithm is as follows:
+--
+-- 1. Try to generate an 'Address' by picking a random index;
+-- 2. If the operation succeeds, return the 'Address';
+-- 3. If the DB operation fails due to a collision, try again, up to a max of
+--    1024 attempts.
+-- 4. If after 1024 attempts there is still no result, flag this upstream.
```

Comment:
It would be good if someone with a good grasp of statistics would do some computation so figure out what the consueqneces of this limit are. I'll ask.